server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNode.java

```diff
@@ -86,21 +91,23 @@ private static Version parseVersion(String version) {
     }
 
     private final Settings settings;
-    private final int processors;
+    private final Processors processors;
```

Comment:
I am not sure this will work in a mixed cluster, since we are changing the type of a field stored in cluster state (regardless of whether the new API has been used) and an old version cannot read it. I think we need a rolling upgrade test. Also, I wonder if we should simply make it two fields instead like `processors` and `processors_range` - to simplify parsing both for ourselves and clients? We could still make the two fields mutually exclusive.

Reply:
I made the assumption that we would be able to break this since we don't have known users of this API, but that might be a risky assumption. I'm wondering if we should prevent upgrades during mixed-version clusters (at least when there are 8.2 and 8.3 master nodes) as we should be able to accept a processor range but the older nodes won't have that notion and we would have to send them the lower bound, maybe it's better to reject the updates in that phase to simplify things? wdty @henningandersen

Reply:
> but that might be a risky assumption

Yes, it can be a bit hard to control that it is never called on an 8.2 cluster. And if so, it will essentially break the cluster during upgrade, which sounds bad.

> we would have to send them the lower bound, maybe it's better to reject the updates in that phase to simplify things?

It sounds reasonable to me to reject a range specification during an upgrade to the release allowing the range specification, since otherwise we would be throwing away data.

I think we need to still accept a single processors specification in a mixed cluster though, i.e., as long as the client conforms to the old version API, everything should still work.